contracts/swaps/SwapsUser.sol

```diff
@@ -160,23 +162,44 @@ contract SwapsUser is State, SwapsEvents, FeesHelper {
 
     function _swapsCall_internal(
         address[5] memory addrs,
-        uint256[3] memory vals)
+        uint256[3] memory vals,
+		bytes memory loanDataBytes)
         internal
         returns (uint256 destTokenAmountReceived, uint256 sourceTokenAmountUsed)
     {
-        bytes memory data = abi.encodeWithSelector(
-            ISwapsImpl(swapsImpl).dexSwap.selector,
-            addrs[0], // sourceToken
-            addrs[1], // destToken
-            addrs[2], // receiverAddress
-            addrs[3], // returnToSenderAddress
-            vals[0],  // minSourceTokenAmount
-            vals[1],  // maxSourceTokenAmount
-            vals[2]   // requiredDestTokenAmount
-        );
+		bytes memory data;
+		address swapImplAddress;
+		if(loanDataBytes.length==0){
+			swapImplAddress = IDexRecords(swapsImpl).retrieveDexAddress(1); //if nothing specified, default to first dex option available. ensure it does not require any input data or else this will break
+			data = abi.encodeWithSelector(
```

Comment:
I think encodeWithSelector can be move out of if else. but check if its worth doing it